rpcserver.go

```diff
@@ -8600,6 +8651,96 @@ func (r *rpcServer) DecodeAssetPayReq(ctx context.Context,
 	return &resp, nil
 }
 
+// syncAssetGroup checks if we already know any asset leaves associated with
+// this group key. If not, it will sync the asset group and return the asset
+// ID of the first leaf found. If there are no leaves found after syncing, an
+// error is returned.
+func (r *rpcServer) syncAssetGroup(ctx context.Context,
+	groupKey *btcec.PublicKey) (asset.ID, error) {
+
+	// We first check if we already know any asset leaves associated with
+	// this group key.
+	leaf := fn.NewRight[asset.ID](*groupKey)
+	leaves, err := r.cfg.Multiverse.FetchLeaves(
+		ctx, []universe.MultiverseLeafDesc{leaf},
+		universe.ProofTypeIssuance,
+	)
+	if err != nil {
+		return asset.ID{}, fmt.Errorf("error fetching leaves: %w", err)
+	}
+
+	tapdLog.Tracef("Found %d leaves for group key %x",
+		len(leaves), groupKey.SerializeCompressed())
+
+	// If there are no leaves, then we need to sync the asset group.
+	if len(leaves) == 0 {
+		tapdLog.Debugf("No leaves found for group key %x, "+
+			"syncing asset group", groupKey.SerializeCompressed())
+		err = r.cfg.AddrBook.SyncAssetGroup(ctx, groupKey)
+		if err != nil {
+			return asset.ID{}, fmt.Errorf("error syncing asset "+
+				"group: %w", err)
+		}
+
+		// Now we can try again.
+		leaves, err = r.cfg.Multiverse.FetchLeaves(
+			ctx, []universe.MultiverseLeafDesc{leaf},
+			universe.ProofTypeIssuance,
+		)
+		if err != nil {
+			return asset.ID{}, fmt.Errorf("error fetching leaves: "+
+				"%w", err)
+		}
+
+		tapdLog.Tracef("Found %d leaves for group key %x",
+			len(leaves), groupKey.SerializeCompressed())
+
+		if len(leaves) == 0 {
+			return asset.ID{}, fmt.Errorf("no asset leaves found "+
+				"for group %x after sync",
+				groupKey.SerializeCompressed())
+		}
+	}
+
+	// Since we know we have at least one leaf, we can just take leaf ID and
+	// query the keys with it. We just need one so we can fetch the actual
+	// proof to find out the asset ID.
+	leafID := leaves[0]
+	leafKeys, err := r.cfg.Multiverse.UniverseLeafKeys(
+		ctx, universe.UniverseLeafKeysQuery{
+			Id:    leafID.ID,
+			Limit: 1,
```

Comment:
What order are we actually relying on here?

Reply:
No specific order (meaning natural database order), since it's not really important which asset we get.